### PR TITLE
Unittests init

### PR DESCRIPTION
**Pull request by jdday**

### CODE REVIEW

File: soil-moisture-sensor/app_class.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def take_readings(soil_moisture_monitor):
+    while True:
+        soil_moisture_monitor.get_reading()
+        time.sleep(10)
+
+
+def main():
+    with open('connection_string.txt', 'r') as file:
```
Comment (CallumTarttelin) [inline soil-moisture-sensor/app_class.py:50]: Just use `os.getEnv` and an environment variable

File: soil-moisture-sensor/app_class.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        device_client = connect(connection_string=file.read())
+    adc = ADC()
+    relay = GroveRelay(5)
+    soil_moisture_monitor = SoilMoistureMonitor(
+        device_client=device_client,
+        adc=adc,
+        relay=relay
+    )
+    if os.getenv('PROD') == 'TRUE':
```
Comment (CallumTarttelin) [inline soil-moisture-sensor/app_class.py:59]: Surely the sensors just fail outside of prod? What is non prod behaviour?

File: soil-moisture-sensor/app_class.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        elif request.name == "relay_off":
+            self.relay.off()
+
+        method_response = MethodResponse.create_from_method_request(request, 200)
+        self.device_client.send_method_response(method_response)
+
+    def get_reading(self):
+        soil_moisture = self.adc.read(0)
+        print("Soil moisture:", soil_moisture)
+        print(dir(Message))
```
Comment (CallumTarttelin) [inline soil-moisture-sensor/app_class.py:30]: weird thing to print

File: soil-moisture-sensor/test_app.py [hunk 4]
```diff
@@ -1,2 +1,64 @@
...
-    assert 1 == 1
+from unittest.mock import MagicMock, patch, mock_open
+
+from app_class import SoilMoistureMonitor, connect, main
+
+
+@patch('app_class.MethodResponse')
+def test_handle_request(mock_method_response):
+    mock_device_client = MagicMock()
+    mock_adc = MagicMock()
```
Comment (CallumTarttelin) [inline soil-moisture-sensor/test_app.py:9]: Consider using spec=...

File: soil-moisture-sensor/test_app.py [hunk 5]
```diff
@@ -1,2 +1,64 @@
...
+    mock_relay = MagicMock()
+    mock_soil_moisture_monitor = SoilMoistureMonitor(adc=mock_adc, device_client=mock_device_client, relay=mock_relay)
+    mock_soil_moisture_monitor.get_reading()
+    mock_adc.read.assert_called_once_with(0)
+    mock_message.assert_called_once_with(mock_json.dumps({'soil_moisture': mock_adc.read.return_value}))
+    mock_device_client.send_message.assert_called_once_with(mock_message.return_value)
+
+
+@patch('app_class.IoTHubDeviceClient')
+def test_connect(mock_iot_hub_device_client):
```
Comment (CallumTarttelin) [inline soil-moisture-sensor/test_app.py:40]: what an unnecessarily long name, `mock_client` should do or something similar

**Review by CallumTarttelin — CHANGES_REQUESTED** [review]

**Comment by CallumTarttelin** [comment]
Also why is `app_class.py` a thing, we should just have one app

File: soil-moisture-sensor/test_app.py [hunk 6]
```diff
@@ -1,2 +1,64 @@
-def test_assert():
-    assert 1 == 1
+from unittest.mock import MagicMock, patch, mock_open
+
+from app_class import SoilMoistureMonitor, connect, main
+
+
+@patch('app_class.MethodResponse')
+def test_handle_request(mock_method_response):
```
[the remark this answers is not in the document]
Comment (jdday) [reply soil-moisture-sensor/test_app.py:7]: -- Split up relay on/off

**Review by CallumTarttelin — CHANGES_REQUESTED** [review]

**Review by CallumTarttelin — APPROVED** [review]

**Pull request merged**